### PR TITLE
Enforce that java.io.tmpdir exists on startup

### PR DESCRIPTION
**Pull request by droberts195**
If the default java.io.tmpdir is used then the startup script creates
it, but if a custom java.io.tmpdir is used then the user must ensure it
exists before running Elasticsearch. If they forget then it can cause
errors that are hard to understand, so this change adds an explicit
check early in the bootstrap and reports a clear error if java.io.tmpdir
is not an accessible directory.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java [hunk 1]
```diff
@@ -288,6 +288,12 @@ static void init(
         } catch (IOException e) {
             throw new BootstrapException(e);
         }
+        // a misconfigured java.io.tmpdir can cause hard-to-diagnose problems later, so reject it immediately
```
Comment (jasontedor) [inline server/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: I am good with the change, but I wonder if we should validate it sooner (e.g., in Elasticsearch). There are other components of the system that might touch the `java.io.tmpdir` before this validation is performed?

**Review by jasontedor — COMMENTED** [review]
I am so sorry for the slow review. I left a comment for consideration.

File: server/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java (hunk 1 above)
Comment (droberts195) [reply server/src/main/java/org/elasticsearch/bootstrap/Bootstrap.java]: Sure, I pushed another commit with the check moved into `Elasticsearch.execute()`.  That method seems to be the earliest one in the `Elasticsearch` class that has access to an `Environment` object.

**Review by jasontedor — APPROVED** [review]
LGTM.

**Pull request merged**